src/frontend/apps/standalone_site/src/components/SortableTable/index.tsx

```diff
@@ -0,0 +1,266 @@
+import { Box, CheckBox, Menu, Pagination, Stack, Text } from 'grommet';
+import { Maybe } from 'lib-common';
+import { Spinner } from 'lib-components';
+import { Fragment, ReactNode, useEffect, useState } from 'react';
+
+interface SortByBase {
+  label: string;
+}
+
+interface SortableTableBase<ItemType> {
+  title: React.ReactNode;
+  loading?: boolean;
+  items: (() => Promise<ItemType[]>) | ItemType[];
+  children: (item: ItemType) => ReactNode;
+}
+
+interface UnSortable {
+  sortable?: false;
+}
+interface Sortable<ItemType, SortByType extends SortByBase = SortByBase> {
+  sortable: true;
+  currentSort: SortByType;
+  sortBy: SortByType[];
+  onSortChange: (newSort: SortByType) => ItemType[] | Promise<ItemType[]>;
+}
+
+interface UnSelectable {
+  selectable?: false;
+}
+interface Selectable<ItemType> {
+  selectable: true;
+  onSelectionChange: (newSelection: ItemType[]) => void;
+}
+
+interface UnPaginable {
+  paginable?: false;
+}
+interface Paginable<ItemType> {
+  paginable: true;
+  currentPage?: number;
+  numberOfItems: number;
+  pageSize?: number;
+  onPageChange: (newPage: number) => ItemType[] | Promise<ItemType[]>;
+}
+
+type SortableTableProps<
+  ItemType,
+  SortByType extends SortByBase = SortByBase,
+> = SortableTableBase<ItemType> &
+  (UnSortable | Sortable<ItemType, SortByType>) &
+  (Selectable<ItemType> | UnSelectable) &
+  (Paginable<ItemType> | UnPaginable);
+
+export const SortableTable = <
+  ItemType,
+  SortByType extends SortByBase = SortByBase,
+>({
+  title,
+  loading: initialLoading,
+  items: initialItems,
+  children,
+  ...props
+}: SortableTableProps<ItemType, SortByType>) => {
+  const [isLoading, setIsLoading] = useState(true);
+  const [items, setItems] = useState<ItemType[]>([]);
+  const [previousSelection, setPreviousSelection] = useState<ItemType[]>([]);
+  const [currentSelection, setCurrentSelection] = useState<ItemType[]>([]);
+  const [targetSort, setTargetSort] = useState<Maybe<SortByType>>(undefined);
+  const [asyncLoader, setAsyncLoader] =
+    useState<Maybe<() => Promise<ItemType[]>>>(undefined);
+
+  let selectionCallback: Maybe<Selectable<ItemType>['onSelectionChange']> =
+    undefined;
+  if (props.selectable) {
+    selectionCallback = props.onSelectionChange;
+  }
+
+  let sortCallback: Maybe<Sortable<ItemType, SortByType>['onSortChange']> =
+    undefined;
+  let SortMenu: Maybe<JSX.Element>;
+  if (props.sortable) {
+    sortCallback = props.onSortChange;
+
+    const sortItems = props.sortBy;
+    const sortCurrentItem = props.currentSort;
+    const availableItems = sortItems.filter((item) => item !== sortCurrentItem);
+
+    SortMenu = (
+      <Menu
+        a11yTitle="Sort item in the table"
+        margin={{ left: 'auto', vertical: 'auto' }}
+        label={sortCurrentItem.label}
+        items={availableItems.map((item) => ({
+          label: item.label,
+          onClick: () => {
+            setTargetSort(item);
+          },
+        }))}
+      />
+    );
+  }
```

Comment:
Should not be after all the `useEffect` ?

Reply:
It can be after all the `useEffect` but I don't understand why it should. (I usually set the effect right before the jsx but if you prefer I do otherwise tel me).

Reply:
Isn't is a react rule to declare all the hooks at the top of the component ?